### PR TITLE
Fix parabolic ray distance calculation in XR teleportation

### LINKED ISSUE #10591: [XR] improve teleportation parabolic ray
The XR parabolic ray's distance calculation is not accurate under certain conditions. 
It is currently constructed from 2 points that are calculated based on the provided radius. 
The praabolic radius should be the max distance between the user and the end point, but it is right now the distance between the user and the first point in the air (from which another ray is sent until it hits the ground).

### PR DESCRIPTION
**Pull request by RaananW**
## Bug Fix: Parabolic Ray Distance Calculation

Fixes #10591

### Problem

The XR teleportation parabolic ray's distance calculation was **not computing distances correctly**. This can be considered a bug — the `parabolicCheckRadius` property was documented as *"the distance from the user to the inspection point"*, but the actual landing point could be **several times further** than the configured radius due to an incorrect compensation formula.

The old code used an arbitrary compensation factor:
```ts
const compensation = 1 + (Math.PI / 2 - Math.abs(xRotation));
const radius = this.parabolicCheckRadius * compensation;
```
This inflated the effective radius by up to ~2.6x and had no geometric relationship to where the ray would actually land. The intermediate "air point" was placed at `radius * 2` forward, making landing distances unpredictable — for example, with the default radius of 5m, pointing at 30° above horizontal could land the user ~18m away.

### Fix

Replaced the compensation-based approach with **horizontal-component scaling** — a physically intuitive model where the direction vector's horizontal magnitude naturally controls reach:

- `effectiveRadius = parabolicCheckRadius * horizontalLength` where `horizontalLength = sqrt(dx^2 + dz^2)`
- Pointing horizontally (`horizontalLength ~ 1`) gives full reach up to `parabolicCheckRadius`
- Pointing steeply upward (`horizontalLength -> 0`) gives near-zero reach (landing very close to the user)
- Pointing almost straight up/down returns early (no parabolic ray computed)

The "air point" is now placed at `effectiveRadius` (not `2x`) and the ray origin at `0.5 * effectiveRadius`, halving the radius-dependent contribution to landing distance.

### Changes

- **Extracted `_buildParabolicRay()`** private method — encapsulates ray construction, makes it testable, and keeps `_onXRFrame` cleaner
- **Removed** the arbitrary `1 + (PI/2 - |xRotation|)` compensation factor
- **Reduced per-frame allocations** — uses `scaleToRef`/`scaleAndAddToRef` instead of `scale()` which created temporary vectors
- **Added 13 unit tests** covering edge cases, direction correctness, scaling behavior, and landing distance properties

### Landing Distance Comparison (controller at 1.5m, radius=5)

| Elevation | Old Distance | New Distance |
|-----------|-------------|-------------|
| 30° up    | ~18m        | ~6.4m       |
| 45° up    | ~10.4m      | ~4m         |
| 60° up    | ~4.7m       | ~2.1m       |
| 80° up    | ~0.6m       | ~0.4m       |

### Backward Compatibility

- No public API changes — `parabolicCheckRadius`, `parabolicRayEnabled`, and all other public properties remain unchanged
- Default value of `parabolicCheckRadius = 5` is preserved
- The behavior change makes the property work as documented

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18154/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18154/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18154/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18154/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18154/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18154/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to the build configuration file.

*This build will release a new package on npm*

If that was unintentional please make sure to revert those changes or close this PR.

**Comment by bjsplat** [comment]
Smart Filters Editor is available to test at:
https://sfe.babylonjs.com/?snapshot=refs/pull/18154/merge

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgl2playwright/index.html

File: packages/dev/core/src/XR/features/WebXRControllerTeleportation.ts [hunk 1]
```diff
@@ -1043,6 +1032,49 @@ export class WebXRMotionControllerTeleportation extends WebXRAbstractFeature {
...
+        // Skip if pointing almost straight up or down — no meaningful horizontal reach
+        if (horizontalLength < 0.01) {
+            return false;
+        }
+
+        // Effective radius scales with horizontal component of direction.
+        // Pointing horizontally (horizontalLength ~ 1) gives full reach.
+        // Pointing up (horizontalLength ~ 0) gives minimal reach.
+        const effectiveRadius = this.parabolicCheckRadius * horizontalLength;
+
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/XR/features/WebXRControllerTeleportation.ts:1061]: This is inherent to parabolic arc geometry — controller height above the floor always contributes additional reach via `origin.y * cot(elevation)`. Clamping to exactly `parabolicCheckRadius` would require solving for the arc-floor intersection with height compensation, which would change the feel of the arc and add complexity without clear benefit. The new behavior is already a massive improvement over the old code (which inflated distances by up to ~2.6x with no geometric basis). The doc comment has been tightened to avoid implying a strict bound (see comment 2 fix below).

File: packages/dev/core/src/XR/features/WebXRControllerTeleportation.ts [hunk 2]
```diff
@@ -1043,6 +1032,49 @@ export class WebXRMotionControllerTeleportation extends WebXRAbstractFeature {
...
+     * Computes the parabolic ray origin and direction for teleportation.
+     * The effective radius scales with the horizontal component of the controller direction,
+     * so pointing more vertically results in shorter landing distances and pointing
+     * horizontally results in landing at approximately `parabolicCheckRadius`.
+     * Modifies the ray in place — the new origin serves as both the raycast start
+     * and the Bezier control point for the visual arc.
+     * @param ray - The controller ray (origin at controller, direction is forward). Modified in place.
+     * @param tmpVector - Pre-allocated temporary vector for computation. Modified in place.
+     * @returns true if the parabolic ray was computed, false if direction is too vertical
+     */
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/XR/features/WebXRControllerTeleportation.ts:1046]: Fixed ✅ — Updated the doc comment to clarify that landing distance also depends on controller height and that `parabolicCheckRadius` controls the forward offset rather than enforcing a strict max ground distance.

File: packages/dev/core/test/unit/XR/webXRControllerTeleportation.test.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        deterministicLockstep: false,
+        lockstepMaxSteps: 1,
+    });
+    const scene = new Scene(engine);
+    const sessionManager = new WebXRSessionManager(scene);
+    const feature = new WebXRMotionControllerTeleportation(sessionManager, {
+        xrInput: { xrCamera: {} } as any,
+    });
+    feature.parabolicCheckRadius = parabolicCheckRadius;
+    return feature;
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/test/unit/XR/webXRControllerTeleportation.test.ts]: Fixed ✅ — `createTeleportationFeature` now returns `{ feature, dispose }` and every call site invokes `dispose()` after assertions. Additionally provided a dummy `teleportationTargetMesh` to skip `_createDefaultTargetMesh` (which requires a canvas 2D context unavailable in jsdom), fixing the pre-existing test suite failure. Used a direct import path for `AbstractMesh` to avoid the draco barrel import issue.

File: packages/dev/core/test/unit/XR/webXRControllerTeleportation.test.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    if (Math.abs(ray.direction.y) < 1e-10) {
+        return null; // ray is horizontal, won't hit floor
+    }
+    const t = -ray.origin.y / ray.direction.y;
+    if (t < 0) {
+        return null; // floor is behind the ray
+    }
+
+    const landingZ = ray.origin.z + t * ray.direction.z;
+    return landingZ; // horizontal distance (ray starts at z=0)
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/test/unit/XR/webXRControllerTeleportation.test.ts]: Fixed ✅ — Changed the return value from `landingZ` to `Math.sqrt(landingX² + landingZ²)` for proper XZ-plane distance. While the existing tests all use X=0 (so the old code happened to produce correct results), the new version is robust for arbitrary directions.

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18154/merge/testResults/webgl2playwright/index.html
